src/hydra-queue-runner/queue-monitor.cc

```diff
@@ -481,71 +478,85 @@ Step::ptr State::createStep(ref<Store> destStore,
         throw PreviousFailure{step};
 
     /* Are all outputs valid? */
+    auto outputHashes = staticOutputHashes(*localStore, *(step->drv));
     bool valid = true;
-    DerivationOutputs missing;
-    for (auto & i : step->drv->outputs)
-        if (!destStore->isValidPath(*i.second.path(*localStore, step->drv->name, i.first))) {
-            valid = false;
-            missing.insert_or_assign(i.first, i.second);
+    std::map<DrvOutput, std::optional<StorePath>> missing;
+    for (auto &[outputName, maybeOutputPath] :
+         step->drv->outputsAndOptPaths(*destStore)) {
+      auto outputHash = outputHashes.at(outputName);
+      if (maybeOutputPath.second) {
+        if (!destStore->isValidPath(*maybeOutputPath.second)) {
+          valid = false;
+          missing.insert({{outputHash, outputName}, maybeOutputPath.second});
         }
+      } else {
+        settings.requireExperimentalFeature(Xp::CaDerivations);
+        if (!destStore->queryRealisation(DrvOutput{outputHash, outputName})) {
+          valid = false;
+          missing.insert({{outputHash, outputName}, std::nullopt});
+        }
+      }
+    }
 
     /* Try to copy the missing paths from the local store or from
        substitutes. */
     if (!missing.empty()) {
 
         size_t avail = 0;
-        for (auto & i : missing) {
-            auto path = i.second.path(*localStore, step->drv->name, i.first);
-            if (/* localStore != destStore && */ localStore->isValidPath(*path))
+        for (auto & [i, maybePath] : missing) {
+            if ((maybePath && localStore->isValidPath(*maybePath)))
                 avail++;
-            else if (useSubstitutes) {
+            else if (settings.isExperimentalFeatureEnabled(Xp::CaDerivations) && localStore->queryRealisation(i)) {
+                maybePath = localStore->queryRealisation(i)->outPath;
+                avail++;
+            } else if (useSubstitutes && maybePath) {
                 SubstitutablePathInfos infos;
-                localStore->querySubstitutablePathInfos({{*path, {}}}, infos);
+                localStore->querySubstitutablePathInfos({{*maybePath, {}}}, infos);
                 if (infos.size() == 1)
                     avail++;
             }
         }
 
         if (missing.size() == avail) {
```

Comment:
It seems that copying dependencies to remote builders does not work as I am seeing error messages in the form

```
Jul 02 08:01:15 nas hydra-queue-runner[765078]: possibly transient failure building ‘/nix/store/r486r6s30ikrqdsb2rrrl5kb1jxhccr9-etc.drv’ on ‘build-pc’: error: some dependencies of '/nix/store/r486r6s30ikrqdsb2rrrl5kb1jxhccr9-etc.drv' are missing
```